src/components/MultiHopTrade/hooks/useReceiveAddress.tsx

```diff
@@ -11,29 +12,31 @@ import {
   selectBuyAsset,
   selectManualReceiveAddress,
 } from 'state/slices/swappersSlice/selectors'
-import { useAppDispatch, useAppSelector } from 'state/store'
+import { useAppSelector } from 'state/store'
 
-const getReceiveAddress = async ({
-  asset,
-  wallet,
-  accountMetadata,
-}: GetReceiveAddressArgs): Promise<string | undefined> => {
-  const { chainId } = fromAssetId(asset.assetId)
-  const { accountType, bip44Params } = accountMetadata
-  const chainAdapter = getChainAdapterManager().get(chainId)
-  if (!(chainAdapter && wallet)) return
-  const { accountNumber } = bip44Params
-  try {
-    return await chainAdapter.getAddress({ wallet, accountNumber, accountType })
-  } catch (e) {
-    console.log(e)
-  }
-}
+export const getReceiveAddress = pMemoize(
+  async ({
+    asset,
+    wallet,
+    accountMetadata,
+  }: GetReceiveAddressArgs): Promise<string | undefined> => {
+    const { chainId } = fromAssetId(asset.assetId)
+    const { accountType, bip44Params } = accountMetadata
+    const chainAdapter = getChainAdapterManager().get(chainId)
+    if (!(chainAdapter && wallet)) return
+    const { accountNumber } = bip44Params
+    const address = await chainAdapter.getAddress({ wallet, accountNumber, accountType })
+    return address
+  },
+  {
+    cacheKey: ([{ asset, accountMetadata }]) =>
+      JSON.stringify({ assetId: asset.assetId, accountMetadata }),
+  },
+)
```

Comment:
This improves perf a *ton* and doesn't re-run this guy so many times - which rugged Ledger as every `wallet.getAddress()` call is an effective device prompt and requires the Ledger to be connected on the right app for that call.

This has regression risks, as we're not invalidating on `wallet` change, since wallet isn't serializable. Perhaps `deviceId` as a property is something we could use to at least invalidate on `deviceId` change?

Reply:
https://github.com/shapeshift/web/pull/5336/commits/18c73c57c643efcbc5b7aeed881c0a17297d47c9